File: src/GameSimulation/spatial-grid.ts
import { Entity } from './components.js';

interface Cell {
  items: Entity[];
  stamp: number;
}

export class SpatialGrid {
  private cells: Cell[]; // flat row-major grid
  private cols: number;
  private rows: number;
  private cellSizeInv: number;
  private cellSize: number;
  private width: number;
  private height: number;
  private stamp: number;

  // Per-query dedupe without Set
  private seenMarks: number[] = [];
  private seenStamp = 1;

  constructor(width: number, height: number, cellSize: number) {
    this.width = width;
    this.height = height;
    this.cellSize = cellSize;
    this.cellSizeInv = 1 / cellSize;
    const [cols, rows] = this.dims(width, height, cellSize);
    this.cols = cols;
    this.rows = rows;
    this.cells = new Array<Cell>(cols * rows);
    for (let i = 0; i < this.cells.length; i++) {
      this.cells[i] = { items: [], stamp: 0 };
    }
    this.stamp = 1;
  }

  setCellSize(cellSize: number): void {
    if (cellSize <= 0) {
      throw new Error('SpatialGrid cell size must be greater than zero.');
    }

    this.cellSize = cellSize;
    this.cellSizeInv = 1 / cellSize;
    const [cols, rows] = this.dims(this.width, this.height, cellSize);
    if (cols !== this.cols || rows !== this.rows) {
      this.cols = cols;
      this.rows = rows;
      this.cells = new Array<Cell>(cols * rows);
      for (let i = 0; i < this.cells.length; i++) {
        this.cells[i] = { items: [], stamp: 0 };
      }
    } else {
      for (let i = 0; i < this.cells.length; i++) {
        this.cells[i].items.length = 0;
        this.cells[i].stamp = 0;
      }
      this.stamp = 1;
    }
    this.seenMarks.length = 0;
    this.seenStamp = 1;
  }

  getCellSize(): number {
    return this.cellSize;
  }

  getTotalCellCount(): number {
    return this.cols * this.rows;
  }

  clear(): void {
    this.stamp++;
    if (this.stamp >= Number.MAX_SAFE_INTEGER) {
      for (let i = 0; i < this.cells.length; i++) {
        this.cells[i].items.length = 0;
        this.cells[i].stamp = 0;
      }
      this.stamp = 1;
    }
  }

  private dims(width: number, height: number, cellSize: number): [number, number] {
    const cols = Math.max(1, Math.ceil(width / cellSize));
    const rows = Math.max(1, Math.ceil(height / cellSize));
    return [cols, rows];
  }

  private idx(col: number, row: number): number {
    return row * this.cols + col;
  }

  private clampIndex(v: number, max: number): number {
    if (v < 0) return 0;
    if (v >= max) return max - 1;
    return v;
  }

  insert(entity: Entity): void {
    const body = entity.body;
    if (!body) return;

    let minCol = Math.floor((body.x - body.radius) * this.cellSizeInv);
    let maxCol = Math.floor((body.x + body.radius) * this.cellSizeInv);
    let minRow = Math.floor((body.y - body.radius) * this.cellSizeInv);
    let maxRow = Math.floor((body.y + body.radius) * this.cellSizeInv);

    minCol = this.clampIndex(minCol, this.cols);
    maxCol = this.clampIndex(maxCol, this.cols);
    minRow = this.clampIndex(minRow, this.rows);
    maxRow = this.clampIndex(maxRow, this.rows);

    for (let row = minRow; row <= maxRow; row++) {
      for (let col = minCol; col <= maxCol; col++) {
        const cell = this.cells[this.idx(col, row)];
        if (cell.stamp !== this.stamp) {
          cell.items.length = 0;
          cell.stamp = this.stamp;
        }
        cell.items.push(entity);
      }
    }
  }

  query(entity: Entity): Entity[] {
    const body = entity.body;
    if (!body) return [];
    return this.getEntitiesInRadius(body.x, body.y, body.radius, entity.id);
  }

  getEntitiesInRadius(
    x: number,
    y: number,
    radius: number,
    ignoreEntityId?: number
  ): Entity[] {
    const result: Entity[] = [];

    // Advance per-query stamp (dedupe) and reset on wrap
    this.seenStamp++;
    if (this.seenStamp >= Number.MAX_SAFE_INTEGER) {
      this.seenMarks.fill(0);
      this.seenStamp = 1;
    }

    const ignoreMark =
      ignoreEntityId === undefined ? undefined : ignoreEntityId >>> 0;

    let minCol = Math.floor((x - radius) * this.cellSizeInv);
    let maxCol = Math.floor((x + radius) * this.cellSizeInv);
    let minRow = Math.floor((y - radius) * this.cellSizeInv);
    let maxRow = Math.floor((y + radius) * this.cellSizeInv);

    minCol = this.clampIndex(minCol, this.cols);
    maxCol = this.clampIndex(maxCol, this.cols);
    minRow = this.clampIndex(minRow, this.rows);
    maxRow = this.clampIndex(maxRow, this.rows);

    for (let row = minRow; row <= maxRow; row++) {
      for (let col = minCol; col <= maxCol; col++) {
        const cell = this.cells[this.idx(col, row)];
        if (cell.stamp !== this.stamp) continue;

        const items = cell.items;
        for (let i = 0; i < items.length; i++) {
          const e = items[i];
          const b = e.body;
          if (!b) continue;

          const id = e.id >>> 0;
          if (ignoreMark !== undefined && id === ignoreMark) {
            continue;
          }

          // Dedupe per query; leave precise distance check to narrow-phase
          if (id >= this.seenMarks.length) {
            this.seenMarks.length = id + 1;
          }
          if (this.seenMarks[id] !== this.seenStamp) {
            this.seenMarks[id] = this.seenStamp;
            result.push(e);
          }
        }
      }
    }

    return result;
  }
}

File: wasm-sim/src/spatial_grid.rs
use ahash::AHashSet;
use smallvec::SmallVec;

pub(crate) struct SpatialGrid {
    // Preallocated 2D grid stored row-major in a flat Vec
    cells: Vec<Cell>,
    cols: i32,
    rows: i32,
    width: f32,
    height: f32,
    cell_size_inv: f32,
    cell_size: f32,
    // Frame stamp used to lazily clear cells
    stamp: u32,
    // Per-query deduplication without HashSet
    seen_marks: Vec<u32>,
    seen_stamp: u32,
}

const CELL_INLINE_CAP: usize = 2;
const QUERY_INLINE_CAP: usize = 2;

struct Cell {
    items: SmallVec<[usize; CELL_INLINE_CAP]>,
    stamp: u32,
}

impl SpatialGrid {
    pub(crate) fn new(width: f32, height: f32, cell_size: f32) -> Self {
        assert!(cell_size > 0.0, "cell size must be positive");
        let (cols, rows) = Self::dims(width, height, cell_size);
        let mut cells = Vec::with_capacity((cols as usize) * (rows as usize));
        for _ in 0..(cols as usize) * (rows as usize) {
            cells.push(Cell {
                items: SmallVec::new(),
                stamp: 0,
            });
        }

        Self {
            cells,
            cols,
            rows,
            width,
            height,
            cell_size_inv: 1.0 / cell_size,
            cell_size,
            stamp: 1,
            seen_marks: Vec::new(),
            seen_stamp: 1,
        }
    }

    pub(crate) fn set_cell_size(&mut self, cell_size: f32) {
        assert!(cell_size > 0.0);
        self.cell_size = cell_size;
        self.cell_size_inv = 1.0 / cell_size;

        // Recompute grid dimensions and reinitialize cells
        let (cols, rows) = Self::dims(self.width, self.height, cell_size);
        if cols != self.cols || rows != self.rows {
            self.cols = cols;
            self.rows = rows;
            self.cells.clear();
            self.cells.reserve((cols as usize) * (rows as usize));
            for _ in 0..(cols as usize) * (rows as usize) {
                self.cells.push(Cell {
                    items: SmallVec::new(),
                    stamp: 0,
                });
            }
        } else {
            // Same grid shape; just reset stamps lazily
            self.stamp = 1;
            for cell in &mut self.cells {
                cell.stamp = 0;
                cell.items.clear();
            }
        }
        // Reset seen data as well
        self.seen_marks.clear();
        self.seen_stamp = 1;
    }

    pub(crate) fn cell_size(&self) -> f32 {
        self.cell_size
    }

    pub(crate) fn clear(&mut self) {
        self.stamp = self.stamp.wrapping_add(1);
        if self.stamp == 0 {
            // Extremely rare wrap-around: fully reset stamps
            self.stamp = 1;
            for cell in &mut self.cells {
                cell.stamp = 0;
                cell.items.clear();
            }
        }
    }

    #[inline]
    fn dims(width: f32, height: f32, cell_size: f32) -> (i32, i32) {
        let cols = (width / cell_size).ceil() as i32;
        let rows = (height / cell_size).ceil() as i32;
        (cols.max(1), rows.max(1))
    }

    #[inline]
    fn clamp_index(v: i32, max: i32) -> i32 {
        if v < 0 { 0 } else if v >= max { max - 1 } else { v }
    }

    #[inline]
    fn to_col(&self, x: f32) -> i32 {
        (x * self.cell_size_inv).floor() as i32
    }

    #[inline]
    fn to_row(&self, y: f32) -> i32 {
        (y * self.cell_size_inv).floor() as i32
    }

    pub(crate) fn insert(&mut self, index: usize, x: f32, y: f32, radius: f32) {
        let mut min_col = self.to_col(x - radius);
        let mut max_col = self.to_col(x + radius);
        let mut min_row = self.to_row(y - radius);
        let mut max_row = self.to_row(y + radius);

        // Clamp to grid bounds
        min_col = Self::clamp_index(min_col, self.cols);
        max_col = Self::clamp_index(max_col, self.cols);
        min_row = Self::clamp_index(min_row, self.rows);
        max_row = Self::clamp_index(max_row, self.rows);

        let cols = self.cols as usize;
        for row in min_row..=max_row {
            let base = (row as usize) * cols;
            for col in min_col..=max_col {
                let idx = base + (col as usize);
                let cell = &mut self.cells[idx];
                if cell.stamp != self.stamp {
                    cell.items.clear();
                    cell.stamp = self.stamp;
                }
                cell.items.push(index);
            }
        }
    }

    pub(crate) fn query(
        &mut self,
        x: f32,
        y: f32,
        radius: f32,
        seen_external: &mut AHashSet<usize>,
    ) -> SmallVec<[usize; QUERY_INLINE_CAP]> {
        seen_external.clear();

        // Advance the per-query stamp and reset marks on wrap
        self.seen_stamp = self.seen_stamp.wrapping_add(1);
        if self.seen_stamp == 0 {
            self.seen_marks.fill(0);
            self.seen_stamp = 1;
        }

        let mut results = SmallVec::<[usize; QUERY_INLINE_CAP]>::new();

        let mut min_col = self.to_col(x - radius);
        let mut max_col = self.to_col(x + radius);
        let mut min_row = self.to_row(y - radius);
        let mut max_row = self.to_row(y + radius);

        // Clamp to grid bounds
        min_col = Self::clamp_index(min_col, self.cols);
        max_col = Self::clamp_index(max_col, self.cols);
        min_row = Self::clamp_index(min_row, self.rows);
        max_row = Self::clamp_index(max_row, self.rows);

        let cols = self.cols as usize;
        for row in min_row..=max_row {
            let base = (row as usize) * cols;
            for col in min_col..=max_col {
                let idx = base + (col as usize);
                let cell = &self.cells[idx];
                if cell.stamp != self.stamp {
                    continue;
                }
                for &idx in &cell.items {
                    if idx < self.seen_marks.len() {
                        if self.seen_marks[idx] != self.seen_stamp {
                            self.seen_marks[idx] = self.seen_stamp;
                            results.push(idx);
                        }
                    } else {
                        // Avoid growing a giant marks array for sparse large indices; use hash set fallback
                        if seen_external.insert(idx) {
                            results.push(idx);
                        }
                    }
                }
            }
        }

        results
    }
}
